tools/twix/src/configuration/keys.rs

```diff
@@ -0,0 +1,233 @@
+use std::{collections::HashMap, fmt};
+
+use eframe::egui::{Key, Modifiers};
+use serde::{
+    de::{self, Deserializer},
+    Deserialize,
+};
+use thiserror::Error;
+
+#[cfg_attr(test, derive(PartialEq))]
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Duplicate modifier `{0}`")]
+    DuplicateModifier(String),
+    #[error("Invalid modifier `{0}`")]
+    InvalidModifier(String),
+    #[error("Invalid key `{0}`")]
+    InvalidKey(String),
+}
+
+#[derive(Debug, Deserialize, Clone, Copy, PartialEq)]
+#[serde(rename_all = "snake_case")]
+pub enum KeybindAction {
+    CloseTab,
+    DuplicateTab,
+    FocusAbove,
+    FocusAddress,
+    FocusBelow,
+    FocusLeft,
+    FocusPanel,
+    FocusRight,
+    NoOp,
+    OpenSplit,
+    Reconnect,
+}
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub struct KeybindTrigger {
+    pub key: Key,
+    pub modifiers: Modifiers,
+}
+
+impl KeybindTrigger {
+    pub fn parse_modifier(value: &&str) -> Result<Modifiers, Error> {
+        match *value {
+            "A" => Ok(Modifiers::ALT),
+            "C" => Ok(Modifiers::CTRL),
+            "S" => Ok(Modifiers::SHIFT),
+            _ => Err(Error::InvalidModifier(String::from(*value))),
+        }
+    }
+
+    pub fn parse(v: &str) -> Result<Self, Error> {
+        let parts = v.split('-').collect::<Vec<_>>();
+
+        let Some((raw_key, raw_modifiers)) = parts.split_last() else {
+            return Err(Error::InvalidKey(v.into()));
+        };
+
+        let is_single_ascii_uppercase_letter =
+            matches!(raw_key.as_bytes(), [letter] if letter.is_ascii_uppercase());
+
+        let Some(key) = Key::from_name(raw_key) else {
+            return Err(Error::InvalidKey(v.into()));
+        };
+
+        let mut modifiers = Modifiers {
+            shift: is_single_ascii_uppercase_letter,
+            ..Modifiers::NONE
+        };
+
+        for raw_modifier in raw_modifiers {
+            let modifier = KeybindTrigger::parse_modifier(raw_modifier)?;
+
+            if modifiers.contains(modifier) {
+                return Err(Error::DuplicateModifier(String::from(*raw_modifier)));
```

Comment:
This may be confusing when doing `S-T`.
But may be fine for now unless you have a great idea how to handle it better.

Reply:
This was intentional, I tried to match helix' behavior. I'm open for a debate on what we want in this case.